### ISSUE TITLE
Non-English Characters

### ISSUE THREAD

**Issue by webfolderio** [lhmouse/nano-win#8]
Thanks for the great port. It works quite well on my Windows 10 except non-English characters.

When i press a non-English character, nano did not print the char.

**Comment by lhmouse** [comment]
It does not work for me either. FFS **ncursesw** (the W variant) seems to want MBCS to be in UTF-8 but native Windows CRT only works with the system's code page which is unlikely UTF-8 (it is `936` on my Simplified Chinese system) so it never worked perfectly.

P.S. You may want to tamper with the configure options to both **nano** and **ncurses** and see whether it works. These are beyond my ken. Sorry.

**Comment by sergeevabc** [comment]
Perhaps someone has found a workdaround to switch & type in non-English?

**Comment by eabase** [comment]
Yep, it totally crashes when changing windows keyboard to another language and type anything.
Need to kill process!

**Comment by webfolderio** [comment]
@lhmouse nano build with mingw64 and it uses msvcrt.dll. msvcrt doesn't support utf-8 and this might be the root cause for Non-English chars. As you might know Windows 10 comes with new C-Runtime (UCRT). [UCRT supports utf-8](https://stackoverflow.com/questions/67848972/differences-between-msvcrt-ucrt-and-vcruntime-libraries) and has better compatible with POSIX.

I didn't try but build nano-win with ucrt based compiler will fix the issue.

There are two ways to compile nan-win with ucrt based compiler.

Option 1: Using MSYS2, requires Windows env. and there is a mingw64 ucrt based dev. env. mingw64 ucrt based doesn't work on Linux.

Option 2: There is an llvm-mingw which supports ucrt based c-runtime and works on both Windows & Linux: https://github.com/mstorsjo/llvm-mingw

What you think about this solution?

Thanks

**Comment by lhmouse** [comment]
New builds are now linked against UCRT, so this should be fixed now.

**Comment by flawrist** [comment]
> New builds are now linked against UCRT, so this should be fixed now.

Actually, it just crashes upon launch now, at least on Windows 7 x64.
Tested binary from nano-win_10789_v9.2-12-g78ce74a6.7z

**Comment by lhmouse** [comment]
I just tried on Windows 7 x64 on Intel i5-7500 and it didn't crash. This reminds me about a hardware issue. Does your CPU support AVX?

**Comment by flawrist** [comment]
> @lhmouse 
Does your CPU support AVX?

No. Supported features here are: CMOV, CMPXCHG8, CX16, FXSR, FXSROPT, LAHF, MMX, NX, OSXSAVE, SSE, SSE2, SSE3, SSSE3, SYSCALL, SYSEE, VMX, X87, XSAVE. 

Why would a console text editor need AVX?!

**Comment by lhmouse** [comment]
> > [@lhmouse](https://github.com/lhmouse)
> > Does your CPU support AVX?
> 
> No. Supported features here are: CMOV, CMPXCHG8, CX16, FXSR, FXSROPT, LAHF, MMX, NX, OSXSAVE, SSE, SSE2, SSE3, SSSE3, SYSCALL, SYSEE, VMX, X87, XSAVE.
> 
> Why would a console text editor need AVX?!

There's no good reason for that. It's the default configuration of my GCC.

I have uploaded a new build with `-march=nocona`: https://files.lhmouse.com/nano-win/nano-win_10796_v9.2-19-ge8a761ce.7z

This is now built with `-Os -flto` so it's much smaller than previous builds.

**Comment by flawrist** [comment]
Alas, it crashes.

**Comment by lhmouse** [comment]
Maybe it's not AVX; my GCC has a default target arch of [x86-64-v2](https://github.com/lhmouse/MINGW-packages/blob/40d5906252ab78ddf1819dac78edd7a2944f591c/mingw-w64-gcc/PKGBUILD#L231) so it may use SSE 4.1 floating-point rounding instructions.

I don't have such an old CPU for testing. I'm sorry but I'm unable to provide support for it.

**Comment by flawrist** [comment]
I do not understand why you are inflating the requirements for a console text editor.
Use `x86-64` as a baseline, no need for `x86-64-v2`, because such an editor never needs SSE4+.
For example, even the latest LAME v4 (MP3 encoder) released in July 2026 works fine without it.

![](https://github.com/user-attachments/assets/8eb41731-71fa-4013-9f1c-d042ef132a44)

**Comment by lhmouse** [comment]
CRT libraries are linked from MSYS2 and are already declared to be Windows 10 only: https://www.msys2.org/news/#2026-02-28-dropping-support-for-windows-81

**Comment by flawrist** [comment]
UCRT works fine on Windows 7, e.g. this OpenSSL v3.6.4 (released on 25 Aug 2026):
https://packages.msys2.org/packages/mingw-w64-ucrt-x86_64-openssl
Or this FAAC encoder v2.1 (released on 16 Aug 2026) also works fine on my end:
https://github.com/knik0/faac/releases/download/faac-2.1/faac-2.1-MSYS2-UCRT64.zip

Just try to compile `nano` with `x86-64` arch.

**Comment by lhmouse** [comment]
DeepSeek V4 Pro:

> ## Result: the entire `nano.exe` contains only **1** SSE4+ instruction
> 
> ```
> 140051958: pblendvb  xmm1, xmm2, xmm0        ; SSE4.1
> ```
> 
> It sits in the static function `__bigint_to_string` (in `.text`, no external symbol, `scl 3`), which is called by `__pformat_int`, `__pformat_emit_efloat`, `__pformat_efloat_decimal`, and others.
> 
> ## How it gets pulled in
> 
> 1. **Source object**: member `lib64_libmingwex_a-mingw_pformat.o` of `libmingwex.a`, i.e. the mingw-w64 CRT's `mingw-w64-crt/stdio/mingw_pformat.c` (from the locally installed package `mingw-w64-ucrt-x86_64-crt 14.0.0.r299.g8c4baed92-1`; the source is in `stdio`).
> 2. **Why it is linked into nano.exe**: nano's uses of `sprintf`/`snprintf`/`fprintf` are redirected to the static `__mingw_sprintf`/`__mingw_vsnprintf`/`__mingw_fprintf` → the `__mingw_pformat` engine (all visible in the exe).
> 3. **How the instruction was produced**: `__bigint_to_string` is a plain-C double-dabble (binary→BCD) loop with no intrinsics and no `#pragma`:
>    ```c
>    buff[i] += (buff[i] > 4) ? 3 : 0;
>    ```
>    GCC's auto-vectorizer turned this per-byte `cond ? a : b` into `pcmpgtb + paddb + pblendvb` (`pblendvb` being the SSE4.1 byte-select).
> 
> ## Root cause
> 
> - Your custom `/ucrt64` GCC is configured with **`--with-arch=x86-64-v2`** (`--with-tune=rocketlake`), so with no `-march` flag it predefines `__SSE4_1__`, `__SSE4_2__`, `__POPCNT__` by default (verified via `gcc -dM -E -`).
> - The CRT package was actually built with `CFLAGS='-g -O2'` (see `src/build-MINGW32/config.log` — it did **not** use the `-march=nocona` default from `/etc/makepkg.conf`), so the vectorizer had SSE4.1 available and chose `pblendvb`.
> - A minimal repro confirms it: the same loop compiles to `pblendvb` with default flags, and degrades to pure SSE2 `pand/pandn/por` when `-mno-sse4.1` is added.
> 
> ## Impact and fix
> 
> Strictly speaking, `nano.exe` now implicitly requires SSE4.1 (present on all x86-64 CPUs since Penryn in 2007, so the practical risk is low). To restore a baseline-SSE2 binary, rebuild the CRT package with `-march=nocona` (or `-mno-sse4.1 -mno-sse4.2 -mno-popcnt`), or build in an environment whose GCC defaults to a lower architecture (e.g. `x86-64`).

**Comment by flawrist** [comment]
Since @lhmouse hesitates to build `nano` with `x86-64` in mind, I set out to find alternatives and am eager to share the following findings with the community — solutions that worked on my laptop with Windows 7 x64  and CPU with no AVX and SSE4 instructions (it's Core 2 Duo [Merom](https://en.wikipedia.org/wiki/Merom_(microprocessor))).

* [nano](https://github.com/Alex313031/nano) v9.1 by @Alex313031

> This is a ported, standalone build of GNU Nano for legacy Windows — running all the way down to Windows NT 4.0 (NT 4.0, 2000, XP, Server 2003, Vista, and 7).

![](https://github.com/user-attachments/assets/b1d98a83-15b9-41f6-a38a-96df63d64d15)

* [nano](https://github.com/chawyehsu/nano-for-windows) v9.2 by @chawyehsu

> * Unicode/UTF-8 support (built against UCRT)
> * Syntax highlighting support
> * nanorc support and system-wide nanorc discovery
> * Spell checker and Formatter support (only external spell checkers)
> * Full keyboard binding support
> * positionlog support (Saved anchors restored when file is reopened, v8.5+)
> * Copy/paste support
> * Mouse support

![](https://github.com/user-attachments/assets/635c3def-3360-420b-b764-a38d03c24231)
